### PR TITLE
add ResettingTimer calculation to type. add tests.

### PR DESCRIPTION
**Pull request by nonsense**

### CODE REVIEW

**Comment by nonsense** [comment]
@holisticode please review when you have time.

File: resetting_timer.go [hunk 1]
```diff
@@ -62,6 +66,16 @@ func (NilResettingTimer) Time(func()) {}
 // Update is a no-op.
 func (NilResettingTimer) Update(time.Duration) {}
 
+// Percentiles panics.
+func (NilResettingTimer) Percentiles([]float64) []int64 {
+	panic("Percentiles called on a NilResettingTimer")
```
Comment (holisticode) [inline resetting_timer.go:71]: Do we really want to call `panic` in these cases? I mean of course it's an obvious issue, but I wonder if a more "grateful" handling might be appropriate?

File: resetting_timer.go [hunk 2]
```diff
@@ -133,7 +160,78 @@ func (*ResettingTimerSnapshot) UpdateSince(time.Time) {
...
+		}
+
+		t.thresholdBoundaries = make([]int64, len(percentiles))
+
+		thresholdBoundary := max
+
+		for i, pct := range percentiles {
+			if count > 1 {
+				var abs float64
+				if pct >= 0 {
```
Comment (holisticode) [inline resetting_timer.go:205]: can pct be < 0?

File: resetting_timer_test.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+			wantP50: 0, wantP95: 0, wantP99: 0,
+			wantMin: 0, wantMax: 0, wantMean: 0,
+		},
+		{
+			values:  []int64{1, 10},
+			start:   0,
+			end:     0,
+			wantP50: 1, wantP95: 10, wantP99: 10,
+			wantMin: 1, wantMax: 10, wantMean: 5.5,
+		},
```
Comment (holisticode) [inline resetting_timer_test.go:61]: Is a test with more than two values too trivial?

File: resetting_timer.go (hunk 1 above)
Comment (nonsense) [reply resetting_timer.go:71]: This is how the approach that the go-metrics library is taking. The methods that you are not supposed to call for given types just panic.

You are only supposed to call `Percentiles` on a snapshot. I am also not mad about the lost type safety, but these types are behind the same interface, and I don't want to introduce more changes than necessary.

File: resetting_timer.go (hunk 2 above)
Comment (nonsense) [reply resetting_timer.go:205]: Yes, if you specify -90, you get a lower bound, and not an upper bound. This is pretty much taken from `statsd`.

File: resetting_timer_test.go (hunk 3 above)
Comment (nonsense) [reply resetting_timer_test.go:61]: I don't understand what you mean. I've added tests for boundaries (0 or 1 values in the collection) to illustrate how the package handles percentiles, mean, etc. with few values, and also added tests for 10, 100 values which is the default case.

**Comment by nonsense** [comment]
I will merge this, so that I can actually vendor it and mark the `go-ethereum` PR ready for review. This code will pass another code review anyways.

Let me know if you have any other questions/concerns and we can iterate on it.

**Pull request merged**